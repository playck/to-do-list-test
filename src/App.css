* {
  box-sizing: border-box;
}

.App {
  width: 100%;
  text-align: center;
}

.app_wrapper {
  max-width: 500px;
  width: 100%;
  margin: 0 auto;
}

.navbar {
  font-size: 2rem;
  padding: 1em;
  background-color: blanchedalmond;
}

.add_form {
  display: flex;
  width: 100%;
  height: 30px;
}

.add_input {
  flex: 1;
  margin-right: 10px;
}

.add_button {
  font-size: 1.3rem;
  height: 100%;
  cursor: pointer;
}

.todo_form {
  padding: 10px;
}

.todo {
  display: flex;
  width: 80%;
  list-style: none;
  padding: 5px;
}

.todo span {
  display: inline-block;
  flex: 1;
  padding-bottom: 3px;
  border-bottom: 1px solid lightgray;
}

.button_wrapper {
  display: flex;
  gap: 5px;
}

.todo button {
  font-size: 1.3rem;
  cursor: pointer;
  border-radius: 50%;
  outline: none;
  border: none;
  width: 30px;
  height: 30px;
}
