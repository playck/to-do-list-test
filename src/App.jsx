import "./App.css";
import Navbar from "./components/navbar";
import TodoForm from "./components/todoForm";

function App({ todoList }) {
  return (
    <div className="App">
      <div className="app_wrapper">
        <Navbar />
        <TodoForm todoList={todoList} />
      </div>
    </div>
  );
}

export default App;
