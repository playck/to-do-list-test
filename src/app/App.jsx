import { useCallback, useState } from "react";
import "./App.css";
import Navbar from "../components/navbar/navbar";
import TodoForm from "../components/todoForm/todoForm";

function App({ presenter }) {
  const [todos, setTodos] = useState(presenter.getTodos());

  const handleAddTodo = useCallback((todo) => {
    presenter.addTodo(todo, setTodos);
  }, []);

  const handleRemoveTodo = useCallback((todo) => {
    presenter.removeTodo(todo, setTodos);
  }, []);

  const handleIncrease = useCallback((todo) => {
    presenter.increaseTodoCount(todo, setTodos);
  }, []);

  const handleDecrease = useCallback((todo) => {
    presenter.decreaseTodoCount(todo, setTodos);
  }, []);

  const resetTodoCount = useCallback(() => {
    presenter.resetAllTodoCount(setTodos);
  });

  return (
    <div className="App">
      <div className="app_wrapper">
        <Navbar />
        <TodoForm
          todos={todos}
          handleAddTodo={handleAddTodo}
          handleRemoveTodo={handleRemoveTodo}
          handleIncrease={handleIncrease}
          handleDecrease={handleDecrease}
          resetTodoCount={resetTodoCount}
        />
      </div>
    </div>
  );
}

export default App;
