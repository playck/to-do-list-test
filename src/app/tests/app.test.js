import TodoPresenter from "../todo_presenter";
import renderer from "react-test-renderer";
import App from "../App";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("App", () => {
  let presenter;
  beforeEach(() => {
    presenter = new TodoPresenter([
      { id: 1, name: "코딩 스터디", count: 0 },
      { id: 2, name: "배그 하기", count: 0 },
      { id: 3, name: "홈트 하기", count: 1 },
    ]);
  });

  it("renders", () => {
    const component = renderer.create(<App presenter={presenter} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("App Component", () => {
    beforeEach(() => {
      render(<App presenter={presenter} />);
    });

    it("adds new todo", () => {
      const newTodo = "Coding";
      const input = screen.getByPlaceholderText("할 일은?");
      const button = screen.getByText("등록");
      userEvent.type(input, newTodo);
      userEvent.click(button);
      const addedTodoName = screen.getAllByTestId("todo-name")[3];
      expect(addedTodoName.innerHTML).toBe(newTodo);
      const addedTodoCount = screen.getAllByTestId("todo-count")[3];
      expect(addedTodoCount.innerHTML).toBe("0");
    });

    it("remove the item", () => {
      const firstLi = screen.getAllByTitle("remove")[0];
      userEvent.click(firstLi);
      const nextLi = screen.getAllByTestId("todo-name")[0];
      expect(nextLi.innerHTML).not.toBe("코딩 스터디");
    });

    it("increases the todo counter", () => {
      const button = screen.getAllByTitle("increase")[0];
      userEvent.click(button);
      const count = screen.getAllByTestId("todo-count")[0];
      expect(count.innerHTML).toBe("1");
    });

    it("decreases the todo counter", () => {
      const button = screen.getAllByTitle("decrease")[2];
      userEvent.click(button);
      const count = screen.getAllByTestId("todo-count")[2];
      expect(count.innerHTML).toBe("0");
    });

    it("resets all todo counters", () => {
      const button = screen.getByText("Reset");
      userEvent.click(button);
      screen.getAllByTestId("todo-count").forEach((count) => {
        expect(count.innerHTML).toBe("0");
      });
    });
  });
});
