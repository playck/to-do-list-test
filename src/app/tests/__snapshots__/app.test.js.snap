// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<div
  className="App"
>
  <div
    className="app_wrapper"
  >
    <nav
      className="navbar"
    >
      <span>
        To-Do-List
      </span>
    </nav>
    <div
      className="todo_form"
    >
      <form
        className="add_form"
        onSubmit={[Function]}
      >
        <input
          className="add_input"
          placeholder="할 일은?"
          type="text"
        />
        <button
          className="add_button"
          style={
            Object {
              "fontSize": "10px",
            }
          }
        >
          등록
        </button>
      </form>
      <ul>
        <li
          className="todo"
        >
          <span
            data-testid="todo-name"
          >
            코딩 스터디
          </span>
          <span
            data-testid="todo-count"
          >
            0
          </span>
          <div
            className="button_wrapper"
          >
            <button
              onClick={[Function]}
              title="increase"
            >
              +
            </button>
            <button
              onClick={[Function]}
              title="decrease"
            >
              -
            </button>
            <button
              onClick={[Function]}
              title="remove"
            >
              X
            </button>
          </div>
        </li>
        <li
          className="todo"
        >
          <span
            data-testid="todo-name"
          >
            배그 하기
          </span>
          <span
            data-testid="todo-count"
          >
            0
          </span>
          <div
            className="button_wrapper"
          >
            <button
              onClick={[Function]}
              title="increase"
            >
              +
            </button>
            <button
              onClick={[Function]}
              title="decrease"
            >
              -
            </button>
            <button
              onClick={[Function]}
              title="remove"
            >
              X
            </button>
          </div>
        </li>
        <li
          className="todo"
        >
          <span
            data-testid="todo-name"
          >
            홈트 하기
          </span>
          <span
            data-testid="todo-count"
          >
            1
          </span>
          <div
            className="button_wrapper"
          >
            <button
              onClick={[Function]}
              title="increase"
            >
              +
            </button>
            <button
              onClick={[Function]}
              title="decrease"
            >
              -
            </button>
            <button
              onClick={[Function]}
              title="remove"
            >
              X
            </button>
          </div>
        </li>
      </ul>
      <button
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
